arch/BPF/BPFDisassembler.c

```diff
@@ -407,52 +411,467 @@ static bool getInstruction(cs_struct *ud, MCInst *MI, bpf_internal *bpf)
-		return decodeALU(ud, MI, bpf);
+		return decodeALU(MI, bpf, is_ebpf);
 	case BPF_CLASS_JMP:
-		return decodeJump(ud, MI, bpf);
+		return decodeJump(MI, bpf, is_ebpf);
 	case BPF_CLASS_RET:
-		/* eBPF doesn't have this class */
-		if (EBPF_MODE(ud))
-			return false;
-		return decodeReturn(ud, MI, bpf);
+	/* case BPF_CLASS_JMP32: */
+		if (is_ebpf)
+			return decodeJump(MI, bpf, is_ebpf);
+		else
+			return decodeReturn(MI, bpf);
 	case BPF_CLASS_MISC:
 	/* case BPF_CLASS_ALU64: */
-		if (EBPF_MODE(ud))
-			return decodeALU(ud, MI, bpf);
+		if (is_ebpf)
+			return decodeALU(MI, bpf, is_ebpf);
+		else
+			return decodeMISC(MI, bpf);
+	}
+
+}
+
+static bpf_insn op2insn_ld_cbpf(const uint16_t opcode)
+{
+// Check for regular load instructions
+#define REG_LOAD_CASE(c) case BPF_SIZE_##c: \
+		if (BPF_CLASS(opcode) == BPF_CLASS_LD) \
+			return BPF_INS_LD##c; \
+		else \
+			return BPF_INS_LDX##c;
+
+	switch (BPF_SIZE(opcode)) {
+	REG_LOAD_CASE(W);
+	REG_LOAD_CASE(H);
+	REG_LOAD_CASE(B);
+	REG_LOAD_CASE(DW);
+	}
+#undef REG_LOAD_CASE
+
+	return BPF_INS_INVALID;
+}
+
+static bpf_insn op2insn_ld_ebpf(const uint16_t opcode)
+{ 
+// Check for packet load instructions
+#define PACKET_LOAD_CASE(c) case BPF_SIZE_##c: \
+		if (BPF_MODE(opcode) == BPF_MODE_ABS) \
+			return BPF_INS_LDABS##c; \
+		else if (BPF_MODE(opcode) == BPF_MODE_IND) \
+			return BPF_INS_LDIND##c; \
+		else \
+			return BPF_INS_INVALID;
+
+	if (BPF_CLASS(opcode) == BPF_CLASS_LD) {
+		switch (BPF_SIZE(opcode)) {
+		PACKET_LOAD_CASE(W);
+		PACKET_LOAD_CASE(H);
+		PACKET_LOAD_CASE(B);
+		}
+	}
+#undef PACKET_LOAD_CASE
+
+	// If it's not a packet load instruction, it must be a regular load instruction
+	return op2insn_ld_cbpf(opcode);
+}
+
+static bpf_insn op2insn_st(const uint16_t opcode, const uint32_t imm)
+{
+	/*
+	 * - BPF_STX | ALU atomic operations | BPF_{W,DW}
+	 * - BPF_STX | Complex atomic operations | BPF_{DW}
+	 * - BPF_ST* | BPF_MEM | BPF_{W,H,B,DW}
+	 */
+
+/* During parsing we already checked to make sure the size is D/DW and 
+ * mode is STX and not ST, so we don't need to check again*/
+#define ALU_CASE_REG(c) case BPF_ALU_##c: \
+		if (BPF_SIZE(opcode) == BPF_SIZE_W) \
+			return BPF_INS_A##c; \
+		else \
+			return BPF_INS_A##c##64;
+
+#define ALU_CASE_FETCH(c) case BPF_ALU_##c | BPF_MODE_FETCH: \
+		if (BPF_SIZE(opcode) == BPF_SIZE_W) \
+			return BPF_INS_AF##c; \
+		else \
+			return BPF_INS_AF##c##64;
+
+#define COMPLEX_CASE(c) case BPF_ATOMIC_##c | BPF_MODE_FETCH: \
+		if (BPF_SIZE(opcode) == BPF_SIZE_DW) \
+			return BPF_INS_A##c##64;
+
+	if (BPF_MODE(opcode) == BPF_MODE_ATOMIC) {
+		switch (imm) {
+		ALU_CASE_REG(ADD);
+		ALU_CASE_REG(OR);
+		ALU_CASE_REG(AND);
+		ALU_CASE_REG(XOR);
+		ALU_CASE_FETCH(ADD);
+		ALU_CASE_FETCH(OR);
+		ALU_CASE_FETCH(AND);
+		ALU_CASE_FETCH(XOR);
+		COMPLEX_CASE(XCHG);
+		COMPLEX_CASE(CMPXCHG);
+		default: // Could only be reached if complex atomic operation is used without fetch modifier, or not as DW  
+			return BPF_INS_INVALID;
```

Comment:
Sorry, I should have asked this clearer. With "illegal" state I mean: "there is no possible byte combination as input, which reaches this point of code".

Invalid byte sequences should still be disassembled without asserts or error messages. They just return "invalid instruction".
In this line I wasn't sure if it can be reached with an invalid byte sequence, or only gets reached if something was messed up in the parsing.

In the first case we don't need an assert. In the latter, we do.